File: vcf2gl.pl
#!/usr/bin/perl
#
# This scripts conversts a vcf file to a simpler format for downstream analysis. I am calling this format genetoype likelihood (gl). The first line lists: the number of individuals and loci. The next line has individual ids. This if followed by on line per SNP that gives the SNP id (scaffold, position) and the phred-scaled genotype likelihoods, three per individual. This version only prints a sub-set of loci with maf > $maf. 
#
# 16vi14, this script now prints the non-reference allele frequency for the retained loci to a file so that they can be used to estimate genotypes
#
# 20vii14, this now combines data from multiple files
#
# USAGE: perl vcf2mpgl.pl maf *.vcf
#

my @line = ();
my $word;
my $nind = 0;
my $nloc = 0;
my $first = 1; ## first vcf file, get ids from here
my $out = "sub_filtered2x_combinedDip.gl";

open (OUT, "> $out") or die "Could not write the outfile\n";

if ($out =~ s/gl/txt/){
	open (OUT2, "> af_$out") or die "Count not write the alt. af file\n";
}

open (OUT3, "> depth_$out") or die "failed to write depth file\n";

my $maf = shift (@ARGV);

foreach my $in (@ARGV){
	open (IN, $in) or die "Could not read the vcf file\n";
	while (<IN>){
		chomp;
		## get individual ids
		if (m/^#CHROM/ & ($first == 1)){
			@line = split(m/\s+/, $_);	
			foreach $word (@line){
				if ($word =~ m/[A-Z]+_[A-Z0-9]+_[DTN]/){
					push (@inds, $word);
					$nind++;
				}
			}
			print OUT "$nind $nloc\n";
			$word = join (" ", @inds);
			print OUT "$word\n";
		}
		## read genetic data lines, write gl
		elsif ((m/^Potrs(\d+)\s+(\d+)/) && (!m/[AGCT],[AGCT]/)){
			$word = "$1".":"."$2";
			if (m/AF=([0-9\.e]+);/){
				$palt = $1;
				if ($palt > 0.5){
					$p = 1 - $palt;
				}
				else {
					$p = $palt;
				}
				if ($p >= $maf){ ## keep this locus
					print OUT "$word";
					@line = split(m/\s+/, $_);
					$i = 0;
					@depth = ();
					foreach $word (@line){
						if ($word =~ s/^\d\/\d\:\d+,\d+\:(\d+)\:\d+\://){
							push(@depth,$1);
							$word =~ s/,/ /g;
							print OUT " $word";
						}
						elsif ($word =~ m/\.\/\./){
							push(@depth,0);
							print OUT " 0 0 0";
						}
				
					}
					print OUT "\n";
					print OUT2 "$palt\n"; ## print p before converting to maf
					$depth = join(" ",@depth);
					print OUT3 "$depth\n";
					$nloc++;
				}
			}
		}
		
	}
}
close (OUT);
close (OUT2);
close (OUT3);
print "Number of loci: $nloc; number of individuals $nind\n";

File: vcf2gl.py
#! /usr/bin/python


# This script reads through a vcf file and extracts (for each SNV) the SNV id, reference allele frequency, alternative allele frequency, the genotype likelihoods for reference, heterozygote and alternative allele, sequence coverage, and the probability that the genotype call is wrong.
# Usage: ./ngs_tools/get_allele_freq_vcf.py <variants.vcf>  > allele_freq.txt



import sys
import re
from collections import OrderedDict


af_file = open(sys.argv[2], 'a')
with open(sys.argv[1], 'rb') as file:
    n_snv = 0
    n_no_num = 0
    n_num = 0
    line_dict = dict()
    for line in file:
        if line[0:2] == '##':
            continue
        elif line[0:2] == '#C':
            header = line.split('\t')
            ind_ids = header[9:len(header)]
            ind_ids[len(ind_ids)-1] = ind_ids[len(ind_ids)-1].split('\n')[0]
        else:
            line_list = line.split('\t')
            if len(line_list[4]) > 1:
                continue
            else:
                geno_likely = OrderedDict()
                n_snv += 1
                n_ind = len(ind_ids)
                scaffold = line_list[0].split('d')[1]
                pos = line_list[1]
                snv_id = ':'.join([scaffold, pos])
                af = re.findall('AF=[0.0-9.0a-z\-]+', line_list[7])
                af = float(af[0].split('=')[1])
                rf = 1 - af
                af_line = str(snv_id + ' ' + str(af) + '\n')
                af_file.write(af_line)
                for j, ind in enumerate(ind_ids):
                    ind_line = str()
                    if line_list[j+9][0:3] == './.':
                        rr, ra, aa = map(str,[0, 0, 0])
                        ind_line = str(rr + ' ' + ra + ' ' + aa)
                        n_no_num += 1
                        geno_likely[ind] = ind_line
                        #print ind, rr, ra, aa
                    else:
                        n_num += 1
                        gt_lklhd = line_list[j+9].split(':')
                        lklhd = gt_lklhd[4].split('\n')[0]
                        rr, ra, aa = lklhd.split(',')
                        #print ind, rr, ra, aa
                        ind_line = str(rr + ' ' + ra + ' ' + aa)
                        geno_likely[ind] = ind_line
                        #print ind_line
                    line_dict[snv_id] = ' '.join(geno_likely.values())
    file.close()
    af_file.close()

print n_ind, n_snv, n_no_num, n_num
print ' '.join(map(str, ind_ids))
it = iter(sorted(line_dict.iteritems()))
for key, value in it:
    print key, value


